Source/EventFlow.PostgreSql/SnapshotStores/PostgresSqlSnapshotPersistence.cs

```diff
@@ -0,0 +1,145 @@
+using EventFlow.Snapshots.Stores;
+
+namespace EventFlow.PostgreSql.SnapshotStores
+{
+    public class PostgreSqlSnapshotPersistence : ISnapshotPersistence
+    {
+        private readonly IPostgreSqlConnection _postgreSqlConnection;
+
+        public PostgreSqlSnapshotPersistence(
+            IPostgreSqlConnection postgreSqlConnection)
+        {
+            _postgreSqlConnection = postgreSqlConnection;
+        }
+
+        public async Task<CommittedSnapshot> GetSnapshotAsync(
+            Type aggregateType,
+            IIdentity identity,
+            CancellationToken cancellationToken)
+        {
+            var postgreSqlSnapshotDataModels = await _postgreSqlConnection.QueryAsync<PostgreSqlSnapshotDataModel>(
+                Label.Named("fetch-snapshot"),
+                cancellationToken,
+                "SELECT TOP 1 * FROM EventFlowSnapshots WHERE AggregateName = @AggregateName AND AggregateId = @AggregateId ORDER BY AggregateSequenceNumber DESC;",
+                new {AggregateId = identity.Value, AggregateName = aggregateType.GetAggregateName().Value})
+                .ConfigureAwait(false);
+
+            if (!postgreSqlSnapshotDataModels.Any())
+            {
+                return null;
+            }
+
+            var postgreSqlSnapshotDataModel = postgreSqlSnapshotDataModels.Single();
+            return new CommittedSnapshot(
+                postgreSqlSnapshotDataModel.Metadata,
+                postgreSqlSnapshotDataModel.Data);
+        }
+
+        public async Task SetSnapshotAsync(
+            Type aggregateType,
+            IIdentity identity,
+            SerializedSnapshot serializedSnapshot,
+            CancellationToken cancellationToken)
+        {
+            var postgreSqlSnapshotDataModel = new PostgreSqlSnapshotDataModel
+            {
+                    AggregateId = identity.Value,
+                    AggregateName = aggregateType.GetAggregateName().Value,
+                    AggregateSequenceNumber = serializedSnapshot.Metadata.AggregateSequenceNumber,
+                    Metadata = serializedSnapshot.SerializedMetadata,
+                    Data = serializedSnapshot.SerializedData,
+                };
+
+            try
+            {
+                await _postgreSqlConnection.ExecuteAsync(
+                    Label.Named("set-snapshot"),
+                    cancellationToken,
+                    @"INSERT INTO EventFlowSnapshots
+                        (AggregateId, AggregateName, AggregateSequenceNumber, Metadata, Data)
+                        VALUES
+                        (@AggregateId, @AggregateName, @AggregateSequenceNumber, @Metadata, @Data);",
+                    postgreSqlSnapshotDataModel)
+                    .ConfigureAwait(false);
+            }
+            catch (SqlException sqlException) when (sqlException.Number == 2601)
+            {
+                // If we have a duplicate key exception, then the snapshot has already been created
```

Comment:
This may also be in the MSSQL/SQLite, but I do think a debug log should be logged with the comment message